File: scala/space-age/src/main/scala/SpaceAge.scala
object SpaceAge {
  private val SecondsInEarthYear: Int = 31557600

  trait Planet {
    def orbitalPeriodOnEarth: Double = orbitalPeriod(this)
  }

  def orbitalPeriod(planet: Planet): Double =
    planet match {
      case Mercury => 0.2408467
      case Venus   => 0.61519726
      case Mars    => 1.8808158
      case Jupiter => 11.862615
      case Saturn  => 29.447498
      case Uranus  => 84.016846
      case Neptune => 164.79132
      case _       => 0.00
    }

  case object Mercury extends Planet
  case object Venus   extends Planet
  case object Mars    extends Planet
  case object Jupiter extends Planet
  case object Saturn  extends Planet
  case object Uranus  extends Planet
  case object Neptune extends Planet

  def ageOn(planet: Planet): Double => Double =
    (seconds: Double) => {
      onEarth(seconds) / planet.orbitalPeriodOnEarth
    }

  def onEarth(seconds: Double): Double   = seconds / SecondsInEarthYear
  def onMercury(seconds: Double): Double = ageOn(Mercury)(seconds)
  def onVenus(seconds: Double): Double   = ageOn(Venus)(seconds)
  def onMars(seconds: Double): Double    = ageOn(Mars)(seconds)
  def onJupiter(seconds: Double): Double = ageOn(Jupiter)(seconds)
  def onSaturn(seconds: Double): Double  = ageOn(Saturn)(seconds)
  def onUranus(seconds: Double): Double  = ageOn(Uranus)(seconds)
  def onNeptune(seconds: Double): Double = ageOn(Neptune)(seconds)
}

File: typescript/space-age/space-age.ts
type Planet = 'earth'
  | 'mercury'
  | 'venus'
  | 'mars'
  | 'jupiter'
  | 'saturn'
  | 'uranus'
  | 'neptune';

type PlanetOrbitalPeriod = Record<Planet, number>;

const SECONDS_IN_EARTH_YEARS = 31557600;

export const planetOrbitalPeriods: PlanetOrbitalPeriod = {
  earth: 1,
  mercury: 0.2408467,
  venus: 0.61519726,
  mars: 1.8808158,
  jupiter: 11.862615,
  saturn: 29.447498,
  uranus: 84.016846,
  neptune: 164.79132
};

interface ISpaceAge {
  ageOnPlanet: (planet: Planet) => number;
}

export class SpaceAge implements ISpaceAge {
  private readonly ageInSeconds: number;

  constructor(ageInSeconds: number) {
    this.ageInSeconds = ageInSeconds;
  }

  public get seconds(): number {
    return this.ageInSeconds;
  }

  public ageOnPlanet = (planet: Planet): number => this.ageOn(planet);

  private ageOn = (planet: Planet): number => (
    Number(((this.seconds / SECONDS_IN_EARTH_YEARS) / planetOrbitalPeriods[planet]).toFixed(2))
  )
}
